### ISSUE TITLE
make surface_camera probe automatically.

### ISSUE THREAD

**Issue by djrscally** [djrscally/miix-510-cameras#6]
For ease-of-use, we really want the bridge module to "just work" with 0 user intervention. The least annoying way that I can see to do this is to slap in a [`request_module()`](https://www.fsl.cs.sunysb.edu/kernel-api/re659.html) into `cio2_pci_probe()` and just let that handle things. Seems like it should work; I'm testing now.

**Comment by djrscally** [comment]
Yep; this seems to work fine.  With the surface-camera driver renamed to "cio2-bridge" and built into the kernel as a module, this patch automatically inits it as part of the `cio2_pci_probe()` - we rely on that driver existing and being loaded anyway so seems sensible:

```patch
diff --git a/drivers/media/pci/intel/ipu3/ipu3-cio2.c b/drivers/media/pci/intel/ipu3/ipu3-cio2.c
index 92f5eadf2c99..0af1fcdcd3cd 100644
--- a/drivers/media/pci/intel/ipu3/ipu3-cio2.c
+++ b/drivers/media/pci/intel/ipu3/ipu3-cio2.c
@@ -1820,6 +1820,12 @@ static int cio2_pci_probe(struct pci_dev *pci_dev,
 
        v4l2_async_notifier_init(&cio2->notifier);
 
+       /* enable the bridge driver to connect sensors */
+       r = request_module("cio2-bridge");
+
+       if (r)
+               pr_err("Failed to enable the cio2_bridge for some reason\n");
+
        /* Register notifier for subdevices we care */
        r = cio2_parse_firmware(cio2);
        if (r)
```

I just need to build my sensor drivers as installed modules to make sure surface_driver/cio2-bridge still links them ok; I suspect we might have to get the module to ret -EPROBE_DEFER or something if the devices have no driver data just to make sure everything gets loaded ok. If it does though, I think that this is the right way to go.

**Comment by djrscally** [comment]
Ah, problem: the `ipu3_cio2` driver is probing not only before the sensor's drivers, but before the i2c subsystem has even built the `i2c_client` for the sensors. Which means [this check](https://github.com/djrscally/miix-510-cameras/blob/master/surface_camera/surface_camera.c#L236) is failed using this method.

Alternate method; perhaps we can use [device links](https://www.kernel.org/doc/html/latest/driver-api/device_link.html) in our sensor drivers to explicitely state that they rely on ipu3_cio2 being loaded, and then add a [`sync_state`](https://www.kernel.org/doc/html/latest/driver-api/driver-model/driver.html#callbacks) call to that driver that loads surface_camera at that stage.

**Comment by djrscally** [comment]
`sync_state` does work; but **not** in the way I expected; it's still called before the device drivers have finished probing. With a significant delay included - between 1 and 2 seconds - things do work completely automatically, which is nice, but it's quite unsatisfying to have the delay.

**Comment by djrscally** [comment]
So; done in a slightly gnarly way. Each sensor has to add a device link between its own device and the cio2's device, with the `DL_FLAG_STATELESS` link set. The `cio2_pci_probe()` function was modified to check if it had any sensors linked to it yet, and if not to defer probing:

```c
void *sensor;

sensor = (struct device *)list_first_entry_of_null(&pci_dev->dev.links.consumers,
                    struct dev_links_info, consumers);

if (!sensor)
    return -EPROBE_DEFER;
```

That prevents the ipu3-cio2 driver from probing until _at least one_ sensor has linked to it. I then moved the call to `request_module()` that loads `cio2-bridge` to the `sync_state` driver callback, which doesn't run until _all_ of the devices that say they're linked to the cio2 have successfully probed - that makes sure the sensors are up and running before allowing the ipu3-cio2 driver to probe, and the cio2-bridge module is initialized as part of that.

Long story short; fully automatic sensor linking.
